### PR TITLE
Use `unittest.mock`

### LINKED ISSUE #3465: Remove mock from test requirements?
I'm learning how to write mock tests, and I noticed things like `import mock` are workarounds to support PY27 and older PY3. Since CuPy now support PY35+ only and `mock` becomes part of the standard Python library, I suppose this line is no longer needed:
https://github.com/cupy/cupy/blob/74dcb4172578a0771e06f4e44b10b5f73f68fb59/setup.py#L39
and all `import mock` can be replaced by `from unittest import mock`?

### PR DESCRIPTION
**Pull request by emcastillo**
Fix #3465 
Thanks @leofang!

### CODE REVIEW

File: tests/cupy_tests/core_tests/fusion_tests/test_function.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import threading
 import unittest
```
Comment (leofang) [inline tests/cupy_tests/core_tests/fusion_tests/test_function.py:3]:
```suggestion
```

File: tests/cupy_tests/cuda_tests/test_compiler.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import pickle
 import unittest
```
Comment (leofang) [inline tests/cupy_tests/cuda_tests/test_compiler.py:3]:
```suggestion
```

File: tests/cupy_tests/cuda_tests/test_profile.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import unittest
```
Comment (leofang) [inline tests/cupy_tests/cuda_tests/test_profile.py:2]:
```suggestion
```

File: tests/cupy_tests/prof_tests/test_range.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import unittest
```
Comment (leofang) [inline tests/cupy_tests/prof_tests/test_range.py:2]:
```suggestion
```

File: tests/cupy_tests/test_init.py [hunk 5]
```diff
@@ -5,7 +5,7 @@
 import tempfile
 import unittest
```
Comment (leofang) [inline tests/cupy_tests/test_init.py:7]:
```suggestion
```

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_construct.py [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import re
 import unittest
 
-import mock
+from unittest import mock
```
Comment (leofang) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_construct.py:4]:
```suggestion
from unittest import mock

```

File: tests/cupyx_tests/test_optimize.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
-import mock
+from unittest import mock
 import pytest
 import tempfile
 import unittest
```
Comment (leofang) [inline tests/cupyx_tests/test_optimize.py:4]:
```suggestion
import tempfile
import unittest
from unittest import mock

import pytest
```

File: tests/cupyx_tests/test_runtime.py [hunk 8]
```diff
@@ -1,6 +1,6 @@
 import unittest
```
Comment (leofang) [inline tests/cupyx_tests/test_runtime.py:2]:
```suggestion
```

**Comment by leofang** [comment]
(Sorry for the commit noise 😅 Should have mentioned that there's a handy, but semi-hidden, "Add suggestion to batch" button that can apply all suggested changes in one commit. It's visible only in the "Files changed" tab, not the "Conversation" tab.)

**Comment by emcastillo** [comment]
Yeah, It was all greyed out and didn't feel like rebasing.

**Comment by emcastillo** [comment]
I just gave you permission to push to my branch so you can clean it if you feel like doing it

**Comment by leofang** [comment]
Hi @emcastillo I don't think the assignment sets up the write permission to me...

**Comment by kmaehashi** [comment]
@emcastillo Push to pull-req's head branch can only be granted to person with write permission to cupy/cupy.

**Comment by kmaehashi** [comment]
I think it's ok to proceed to merge.

pfnCI, test this please.

**Comment by kmaehashi** [comment]
@emcastillo Could you check test failures?

**Comment by leofang** [comment]
@kmaehashi, @emcastillo gave me permission so I took the liberty to cleaned up the cluttered commits and fix the CI error.

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Pull request merged**